test/e2e/uninstall/uninstall_test.go

```diff
@@ -42,48 +30,69 @@ func deleteDeployYaml(filename string) {
 
-// Encapsulate the logic to find where the templated yamls are
-func getTemplateYamlDir() string {
-	_, filename, _, _ := runtime.Caller(0)
-	return path.Clean(filename + "/../../install/deploy")
-}
-
-func makeImageName(registryAddress string, registryport string, imagename string, imageversion string) string {
-	return registryAddress + ":" + registryport + "/mayadata/" + imagename + ":" + imageversion
-}
-
-// We expect this to fail a few times before it succeeds,
-// so no throwing errors from here.
-func mayastorReadyPodCount() int {
-	var mayastorDaemonSet appsv1.DaemonSet
-	if k8sClient.Get(context.TODO(), types.NamespacedName{Name: "mayastor", Namespace: "mayastor"}, &mayastorDaemonSet) != nil {
-		return -1
-	}
-	return int(mayastorDaemonSet.Status.CurrentNumberScheduled)
+// Helper for deleting mayastor CRDs
+func deleteCRD(crdName string) {
+	cmd := exec.Command("kubectl", "delete", "crd", crdName)
+	_, err := cmd.CombinedOutput()
+	Expect(err).ToNot(HaveOccurred())
 }
 
 // Teardown mayastor on the cluster under test.
 // We deliberately call out to kubectl, rather than constructing the client-go
 // objects, so that we can verfiy the local deploy yamls are correct.
 func teardownMayastor() {
-	deleteDeployYaml("mayastor-daemonset.yaml")
-	deleteDeployYaml("moac-deployment.yaml")
-	deleteDeployYaml("csi-daemonset.yaml")
-	deleteDeployYaml("nats-deployment.yaml")
+	// The correct sequence for a reusable  cluster is
+	// Delete all pods in the default namespace
+	// Delete all pvcs
+	// Then uninstall mayastor
+	podsDeleted, podCount := common.DeleteAllPods()
+	pvcsDeleted, pvcsFound := common.DeleteAllPVCs()
+
+	logf.Log.Info("Cleanup done, Uninstalling mayastor")
+	// Deletes can stall indefinitely, try to mitigate this
+	// by running the deletes in different threads
+	go deleteDeployYaml("mayastor-daemonset.yaml")
+	go deleteDeployYaml("moac-deployment.yaml")
+	go deleteDeployYaml("csi-daemonset.yaml")
+	go deleteDeployYaml("nats-deployment.yaml")
+
+	{
+		iters := 18
+		logf.Log.Info("Waiting for Mayastor pods to be deleted", "timeout seconds", iters*10)
+		numMayastorPods := common.MayastorUndeletedPodCount()
+		for attempts := 0; attempts < iters && numMayastorPods != 0; attempts++ {
+			time.Sleep(10 * time.Second)
+			numMayastorPods = common.MayastorUndeletedPodCount()
+			logf.Log.Info("", "numMayastorPods", numMayastorPods)
+		}
+	}
+
+	// The focus is on trying to make the cluster reusable, so we try to delete everything.
+	// TODO: When we start using a cluster for a single test run  move these set of deletes to after all checks.
 	deleteDeployYaml("mayastorpoolcrd.yaml")
 	deleteDeployYaml("moac-rbac.yaml")
 	deleteDeployYaml("storage-class.yaml")
+	deleteCRD("mayastornodes.openebs.io")
+	deleteCRD("mayastorvolumes.openebs.io")
+	// Attempt to forcefully delete pods
+	// TODO replace this function call when a single cluster is used for a single test run, with a check.
+	forceDeleted := common.ForceDeleteMayastorPods()
```

Comment:
Would there be any benefit in having this after line 87? So we could check if the "standard" delete was sufficient to delete all Mayastor pods. If not, then we force delete and check that this is successful.

Reply:
I did consider that, but for now I think this is "safer", I haven't been able to identify the root cause, but we have encountered situation where the cluster is unusable because namespace is stuck in "terminating". So my approach is to delete all resources and then delete the namspace.